libraries/AP_MSP/AP_MSP_Telem_Backend.cpp

```diff
@@ -0,0 +1,945 @@
+#if OSD_ENABLED
+    units = osd->units == AP_OSD::UNITS_METRIC ? OSD_UNIT_METRIC : OSD_UNIT_IMPERIAL;
+#endif
+
+    sbuf_write_u8(dst, units);                                 // units
+    // Alarms
+    sbuf_write_u8(dst, msp->_osd_config.rssi_alarm);                 // rssi alarm
+    sbuf_write_u16(dst, msp->_osd_config.cap_alarm);                 // capacity alarm
+    // Reuse old timer alarm (U16) as OSD_ITEM_COUNT
+    sbuf_write_u8(dst, 0);
+    sbuf_write_u8(dst, OSD_ITEM_COUNT);                             // osd items count
+
+    sbuf_write_u16(dst, msp->_osd_config.alt_alarm);                 // altitude alarm
+
+    // element position and visibility
+    uint16_t pos = 0;   // default is hide this element
+    for (uint8_t i = 0; i < OSD_ITEM_COUNT; i++) {
+        pos = 0;    // 0 is hide this item
+        if (msp->_osd_item_settings[i] != nullptr) {      // ok supported
+            if (msp->_osd_item_settings[i]->enabled) {    // ok enabled
+                // let's check if we need to hide this dynamically
+                if (!BIT_IS_SET(osd_hidden_items_bitmask, i)) {
+                    pos = MSP_OSD_POS(msp->_osd_item_settings[i]);
+                }
+            }
+        }
+        sbuf_write_u16(dst, pos);
+    }
+
+    // post flight statistics
+    sbuf_write_u8(dst, OSD_STAT_COUNT);                         // stats items count
+    for (uint8_t i = 0; i < OSD_STAT_COUNT; i++ ) {
+        sbuf_write_u16(dst, 0);                                 // stats not supported
+    }
+
+    // timers
+    sbuf_write_u8(dst, OSD_TIMER_COUNT);                      // timers
+    for (uint8_t i = 0; i < OSD_TIMER_COUNT; i++) {
+        // no timer support
+        sbuf_write_u16(dst, 0);
+    }
+
+    // Enabled warnings
+    // API < 1.41
+    // Send low word first for backwards compatibility
+    sbuf_write_u16(dst, (uint16_t)(msp->_osd_config.enabled_warnings & 0xFFFF)); // Enabled warnings
+    // API >= 1.41
+    // Send the warnings count and 32bit enabled warnings flags.
+    // Add currently active OSD profile (0 indicates OSD profiles not available).
+    // Add OSD stick overlay mode (0 indicates OSD stick overlay not available).
+    sbuf_write_u8(dst, OSD_WARNING_COUNT);            // warning count
+    sbuf_write_u32(dst, msp->_osd_config.enabled_warnings);  // enabled warning
+
+    // If the feature is not available there is only 1 profile and it's always selected
+    sbuf_write_u8(dst, 1);    // available profiles
+    sbuf_write_u8(dst, 1);    // selected profile
+
+    sbuf_write_u8(dst, 0);    // OSD stick overlay
+
+    // API >= 1.43
+    // Add the camera frame element width/height
+    //sbuf_write_u8(dst, osdConfig()->camera_frame_width);
+    //sbuf_write_u8(dst, osdConfig()->camera_frame_height);
+    return MSP_RESULT_ACK;
+}
+
+MSPCommandResult AP_MSP_Telem_Backend::msp_process_out_attitude(sbuf_t *dst)
+{
+    AP_AHRS &ahrs = AP::ahrs();
+    WITH_SEMAPHORE(ahrs.get_semaphore());
```

Comment:
Do I really need a semaphore for accessing int32_t?

Reply:
no, but if you want the 3 values to be consistent with each other you do